test/e2e/source_apiserver_test.go

```diff
@@ -82,37 +89,47 @@ func (test *e2eTest) apiServerSourceDelete(t *testing.T, sourceName string) {
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(saCmd, " "), err.Error())
 	}
-	_, err = kubectl.RunWithOpts([]string{"create", "clusterrole", "testsa-role", "--verb=get,list,watch", "--resource=events,namespaces"}, runOpts{})
+
+	crCmd := []string{"create", "clusterrole", clusterRolePrefix + test.kn.namespace, "--verb=get,list,watch", "--resource=events,namespaces"}
+	_, err = kubectl.RunWithOpts(crCmd, runOpts{})
 	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl create clusterrole testsa-role'. Error: %s", err.Error()))
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(crCmd, " "), err.Error())
 	}
-	_, err = kubectl.RunWithOpts([]string{"create", "clusterrolebinding", "testsa-binding", "--clusterrole=testsa-role", "--serviceaccount=" + test.kn.namespace + ":" + name}, runOpts{})
+
+	crbCmd := []string{"create", "clusterrolebinding", clusterRoleBindingPrefix + test.kn.namespace, "--clusterrole=" + clusterRolePrefix + test.kn.namespace, "--serviceaccount=" + test.kn.namespace + ":" + testServiceAccount}
+	_, err = kubectl.RunWithOpts(crbCmd, runOpts{})
 	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl create clusterrolebinding testsa-binding'. Error: %s", err.Error()))
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(crbCmd, " "), err.Error())
 	}
 }
 
-func (test *e2eTest) deleteServiceAccountForApiserver(t *testing.T, name string) {
+func (test *e2eTest) tearDownForSourceApiServer(t *testing.T) {
 	kubectl := kubectl{t, Logger{}}
 
-	_, err := kubectl.RunWithOpts([]string{"delete", "serviceaccount", name, "--namespace", test.kn.namespace}, runOpts{})
+	saCmd := []string{"delete", "serviceaccount", testServiceAccount, "--namespace", test.kn.namespace}
+	_, err := kubectl.RunWithOpts(saCmd, runOpts{})
 	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl delete serviceaccount test-sa'. Error: %s", err.Error()))
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(saCmd, " "), err.Error())
 	}
-	_, err = kubectl.RunWithOpts([]string{"delete", "clusterrole", "testsa-role"}, runOpts{})
+
+	crCmd := []string{"delete", "clusterrole", clusterRolePrefix + test.kn.namespace}
+	_, err = kubectl.RunWithOpts(crCmd, runOpts{})
 	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl delete clusterrole testsa-role'. Error: %s", err.Error()))
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(crCmd, " "), err.Error())
 	}
-	_, err = kubectl.RunWithOpts([]string{"delete", "clusterrolebinding", "testsa-binding"}, runOpts{})
+
+	crbCmd := []string{"delete", "clusterrolebinding", clusterRoleBindingPrefix + test.kn.namespace}
+	_, err = kubectl.RunWithOpts(crbCmd, runOpts{})
 	if err != nil {
-		t.Fatalf(fmt.Sprintf("Error executing 'kubectl delete clusterrolebinding testsa-binding'. Error: %s", err.Error()))
+		t.Fatalf("Error executing '%s'. Error: %s", strings.Join(crbCmd, " "), err.Error())
```

Comment:
```suggestion
		t.Fatalf("Error executing '%s'. Error: %v", strings.Join(crbCmd, " "), err.Error())
```

Reply:
I'm not so deep in the code (and lacking a bit the context here), but how it is ensure, that this delete command is issued if when the test before failed ? Otherwise the next test in this namespace will fail, too.

Maybe we could use `defer` for this ? (like in `defer deleteClusterRoleBinding`, or with a closure when argumentsa re required

Reply:
The namespace is not deleted if any test in it fails, this helps dumping logs in k8s dump at the end of PR job, if namespace is present.

Also, if test fails, its namespace is kept and a subsequent test will not re-use same namespace but create the next in index `kne2etestsN+1`.
This way the cluster wide resources will have namespace string in it and won't collide.


> Maybe we could use defer for this ? (like in defer deleteClusterRoleBinding, or with a closure when argumentsa re required

We're removing SA, CR and CRB in ApiServer source specific tearDown in line no:40
```
  defer func() {
    test.tearDownForSourceApiServer(t)
    test.Teardown(t)
  }() 
```
Do you mean, separating each k8s reource `delete` operation in a method?

Reply:
Yeah, that sounds fair enough. As mentioned, I'm not so in the weeds with the test setup yet.